### PR TITLE
update how_tos/reading_data to use Dataset API

### PR DESCRIPTION
**Pull request by FirefoxMetzger**
Since the Dataset API moved from .contrib.data into .data (core) update the MNIST example to use Dataset over queues.

This was the first page I've found when searching how to best get data into TF and I thought it should reflect the current best practice. Also, if I understand correctly, queues might be deprecated in favor of Datasets some time in the future.

### CODE REVIEW

**Comment by FirefoxMetzger** [comment]
Any thoughts on this PR? Should I write a separate how to rather than updating the existing one?

**Comment by mrry** [comment]
IMO the changes look great, but I will defer to my colleague @MarkDaoust, who has been revising these tutorials and might have stronger opinions.

File: tensorflow/examples/how_tos/reading_data/fully_connected_reader.py [hunk 1]
```diff
@@ -143,47 +149,33 @@ def run_training():
...
-        # the list passed to sess.run() and the value tensors
-        # will be returned in the tuple from the call.
-        _, loss_value = sess.run([train_op, loss])
-
-        duration = time.time() - start_time
-
-        # Print an overview fairly often.
-        if step % 100 == 0:
-          print('Step %d: loss = %.2f (%.3f sec)' % (step, loss_value,
+    with tf.Session().as_default() as sess:
```
Comment (MarkDaoust) [inline tensorflow/examples/how_tos/reading_data/fully_connected_reader.py]: Optional: this `as_default()` is unnecessary.

**Review by MarkDaoust — DISMISSED** [review]
It looks like this docs paragraph would need a small update to account for this change (mostly just removing the link to this file):
 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/docs_src/api_guides/python/reading_data.md#standard-tensorflow-format

File: tensorflow/examples/how_tos/reading_data/fully_connected_reader.py (hunk 1 above)
Comment (mrry) [reply tensorflow/examples/how_tos/reading_data/fully_connected_reader.py]: Actually, I'd upgrade this to "Mandatory:", since the only difference in adding `.as_default()` is to delay closing the `Session` until it is garbage collected. In this code, it might not make a difference, but if somebody copies and pastes it (as they are likely to do) they might find that the session stays alive for longer than they'd like.

**Comment by MarkDaoust** [comment]
I agree with @mrry, This is a great change.

Just note that there is a long-term plan to eventually merge most of the examples directory into tensorflow/models, and some of the code in this sub-directory might not survive the transition. 

So while I encourage you to continue sending PRs, you might want to avoid this sub-directory.

Thanks.

File: tensorflow/examples/how_tos/reading_data/fully_connected_reader.py (hunk 1 above)
Comment (FirefoxMetzger) [reply tensorflow/examples/how_tos/reading_data/fully_connected_reader.py]: Oh, I wasn't even aware of that behavior regarding garbage collection. I've picked it up after carelessly calling `eval()` on a tensor once and the error suggested:

    No default session is registered. Use `with sess.as_default()` or pass an explicit session to 
    `eval(session=sess)`

I'll change that ... here and in my own code. Very good catch, thanks.

File: tensorflow/examples/how_tos/reading_data/fully_connected_reader.py [hunk 2]
```diff
@@ -45,11 +45,9 @@
 VALIDATION_FILE = 'validation.tfrecords'
 
 
-def read_and_decode(filename_queue):
-  reader = tf.TFRecordReader()
-  _, serialized_example = reader.read(filename_queue)
-  features = tf.parse_single_example(
-      serialized_example,
+def decode(serialized_example):
+  features = tf.parse_example(
```
Comment (mrry) [inline tensorflow/examples/how_tos/reading_data/fully_connected_reader.py]: IMO, this would be a bit simpler as `features = tf.parse_single_example(serialized_example, ...)` rather than using the batch version and explicitly adding a dimension to `serialized_example`. You wouldn't need the `tf.reshape(image, [-1])` below either (or the later reshape of the label value), and the comment about "scalar string tensor" would remain correct (in the current PR, it's a length-1 vector of strings).

Did you have a reason to prefer switching it here?

Comment (FirefoxMetzger) [reply tensorflow/examples/how_tos/reading_data/fully_connected_reader.py]: I stumbled upon #390 . I wasn't sure if it was still relevant so I tried both versions and on my machine `parse_example` was slightly faster. Another thought was to do `dataset = dataset.batch(...)` before the parsing. Although, I haven't tested it so can't say if it is better.

If it is easier to read via `tf.parse_single_example` I'll happily revert the change. After all the focus here is readability over performance.

Comment (mrry) [reply tensorflow/examples/how_tos/reading_data/fully_connected_reader.py]: I'd prefer the `parse_single_example` version for readability, please.

(I am curious at what the cause of different performance might be... applying `tf.parse_example()` to a vector containing one example follows almost exactly the same code path as applying `tf.parse_single_example()` to a scalar, except it does all of the necessary squeezing etc. internally. The way it deals with `tf.SparseTensor` appears to be [a bit clumsy](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/parsing_ops.py#L839-L847), but that path isn't used here, and apart from that I don't see any obvious deficiencies....)

Comment (FirefoxMetzger) [reply tensorflow/examples/how_tos/reading_data/fully_connected_reader.py]: Okay, I've just moved batching and shuffling up to try decoding examples as batches rather than single examples. That does make a huge difference in speed and removes both reshape ops. I will still revert the changes and update the PR.

(Just FYI here are the times on my super recent GTX750Ti)

Decoding `parse_single_example`
    Step 0: loss = 2.31 (0.609 sec)
    Step 100: loss = 2.15 (0.031 sec)
    Step 200: loss = 1.96 (0.030 sec)
    Step 300: loss = 1.60 (0.029 sec)
    Step 400: loss = 1.35 (0.029 sec)
    Step 500: loss = 0.90 (0.031 sec)
    Step 600: loss = 0.88 (0.030 sec)
    Step 700: loss = 0.72 (0.031 sec)
    Step 800: loss = 0.62 (0.028 sec)
    Step 900: loss = 0.60 (0.031 sec)
    Step 1000: loss = 0.56 (0.027 sec)
    Done training for 2 epochs, 1100 steps.


Decoding `parse_example` (one example at a time):

    Step 0: loss = 2.30 (0.585 sec)
    Step 100: loss = 2.18 (0.031 sec)
    Step 200: loss = 1.99 (0.031 sec)
    Step 300: loss = 1.70 (0.031 sec)
    Step 400: loss = 1.39 (0.030 sec)
    Step 500: loss = 1.14 (0.029 sec)
    Step 600: loss = 0.73 (0.031 sec)
    Step 700: loss = 0.73 (0.028 sec)
    Step 800: loss = 0.75 (0.030 sec)
    Step 900: loss = 0.59 (0.028 sec)
    Step 1000: loss = 0.80 (0.030 sec)
    Done training for 2 epochs, 1100 steps.

Decoding `parse_example` batched:

    Step 0: loss = 2.35 (0.237 sec)
    Step 100: loss = 2.10 (0.004 sec)
    Step 200: loss = 1.79 (0.004 sec)
    Step 300: loss = 1.42 (0.004 sec)
    Step 400: loss = 1.14 (0.005 sec)
    Step 500: loss = 0.87 (0.003 sec)
    Step 600: loss = 0.64 (0.004 sec)
    Step 700: loss = 0.69 (0.004 sec)
    Step 800: loss = 0.75 (0.005 sec)
    Step 900: loss = 0.45 (0.004 sec)
    Step 1000: loss = 0.57 (0.005 sec)
    Done training for 2 epochs, 1100 steps.

Comment (mrry) [reply tensorflow/examples/how_tos/reading_data/fully_connected_reader.py]: Wow, those numbers are concerning. It looks like there's a decent amount of low-hanging performance to be picked up in the `tf.parse_single_example()` implementation, so I'm going to prioritize that. 

Thanks for making the changes!

**Review by mrry — APPROVED** [review]

**Comment by FirefoxMetzger** [comment]
@MarkDaoust What do you think about the docs update in this PR? Do you prefer it like this or rather have the link deleted completely?
I personally find it a bit odd to declare a "standard format" and than not provide at least one full implementation example, thats why I didn't simply remove the link.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Review by MarkDaoust — APPROVED** [review]
The goal of that section of the doc is to describe the queue based version.  

So added a snippet of code showing how to do it with queues. But I kept the link the updated example saying "this is how you should do it".

Thanks again for the change.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by FirefoxMetzger** [comment]
Looks like the tests keep failing, not just for me but across all PRs and it doesn't seem to be code I've changed. Would it help if I merge the latest master into this PR?

**Comment by caisq** [comment]
@tensorflow-jenkins test this please
